src/Promitor.Core.Scraping/Configuration/Providers/MetricsDeclarationProvider.cs

```diff
@@ -42,7 +43,7 @@ public virtual MetricsDeclaration Get(bool applyDefaults = false)
                     }
                     if (metric.AzureMetricConfiguration?.Aggregation.Interval == null)
```

Comment:
Sure.

Reply:
Actually - thinking about it again that doesn't make sense. If we did that and `Aggregation` was null, the next line would fail with a NullReferenceException.

Reply:
I was actually confusing it with `config.MetricDefaults.Aggregation.Interval`.

This makes me wonder, shouldn't the if check on `config.MetricDefaults.Aggregation.Interval` rather than the metric or what are we trying to achieve here 🤔

Reply:
I think the check makes sense. What it's saying is "if the metric doesn't have an interval defined, use the global interval instead". It's similar to the way that the resource group override logic used to work.

An alternative approach would be to make the runtime model immutable, and then alter the places where we use the aggregation interval to do this check instead, similar to what I did in https://github.com/tomkerkhove/promitor/blob/master/src/Promitor.Core.Scraping/Configuration/Model/Metrics/MetricDefinition.cs#L62. However, I don't think it's worth doing that as part of this set of changes since the code was already working fine.

The reason I had to add the null propagation operator here is that I took a slightly different approach to the config deserialization than had been previously taken. Before my changes, if the default aggregation interval wasn't specified in the yaml, we were deserializing it to an empty `Aggregation` object and setting a default. In the new deserialization code I've written, I've taken the approach of leaving objects that aren't included in the yaml null so that the config objects more closely match the yaml they were deserialized from.

I hadn't brought this up yet because I didn't want to confuse things, but I think it might be worth thinking about taking a slightly different approach to validating the config, or maybe having an additional validation stage. At the moment we run the validation against the runtime model *after* the point it's deserialized, but I think we could actually perform validation checks *during* the deserialization process to produce better information to users.

My rough idea is that because the deserialization code has access to the `YamlNode` objects, we can probably get extra info like the line numbers so we could produce messages like this:

```text
metrics-config.yaml(1, 10): "v3" is not a valid configuration version

version: v3
         ~~
```

I realise that's out of scope for these changes, but just wanted to mention it while I'm thinking about it.

Reply:
> I think the check makes sense. What it's saying is "if the metric doesn't have an interval defined, use the global interval instead". It's similar to the way that the resource group override logic used to work.
> 
> An alternative approach would be to make the runtime model immutable, and then alter the places where we use the aggregation interval to do this check instead, similar to what I did in [/src/Promitor.Core.Scraping/Configuration/Model/Metrics/MetricDefinition.cs@`master`#L62](https://github.com/tomkerkhove/promitor/blob/master/src/Promitor.Core.Scraping/Configuration/Model/Metrics/MetricDefinition.cs#L62) [](/tomkerkhove/promitor/blob/HEAD@{2019-08-24T06:34:27Z}/src/Promitor.Core.Scraping/Configuration/Model/Metrics/MetricDefinition.cs#L62). However, I don't think it's worth doing that as part of this set of changes since the code was already working fine.

Ha that actually makes sense, let's leave it as is!

> The reason I had to add the null propagation operator here is that I took a slightly different approach to the config deserialization than had been previously taken. Before my changes, if the default aggregation interval wasn't specified in the yaml, we were deserializing it to an empty `Aggregation` object and setting a default. In the new deserialization code I've written, I've taken the approach of leaving objects that aren't included in the yaml null so that the config objects more closely match the yaml they were deserialized from.
> 
> I hadn't brought this up yet because I didn't want to confuse things, but I think it might be worth thinking about taking a slightly different approach to validating the config, or maybe having an additional validation stage. At the moment we run the validation against the runtime model _after_ the point it's deserialized, but I think we could actually perform validation checks _during_ the deserialization process to produce better information to users.
> 
> My rough idea is that because the deserialization code has access to the `YamlNode` objects, we can probably get extra info like the line numbers so we could produce messages like this:
> 
> ```
> metrics-config.yaml(1, 10): "v3" is not a valid configuration version
> 
> version: v3
>          ~~
> ```
> 
> I realise that's out of scope for these changes, but just wanted to mention it while I'm thinking about it.

That's a good point, certainly given we support hot-swapping configurations for runtime (which will come to metric declaration as well #573)

Will add a note there to use your suggestion, thanks!